Rework RBAC documentation

israel-hdez opened the pull request:
Since Kiali is no longer allowing to turn off the kube cache, the RBAC documentation no longer stands. It was anyway due an update.

This is a re-write of the RBAC documentation to reflect current behavior.

Fixes kiali/kiali#5720

israel-hdez committed:
Rework RBAC documentation

Since Kiali is no longer allowing to turn off the kube cache, the RBAC
documentation no longer stands. It was anyway due an update.

This is a re-write of the RBAC documentation to reflect current
behavior.

Fixes kiali/kiali#5720

jmazzitelli commented on content/en/docs/Configuration/rbac.md:
````diff
@@ -1,120 +1,138 @@
-$ oc adm policy add-role-to-user kiali john -n mynamespace # For OpenShift clusters
-```
+{{% alert color="warning" %}}
+Kiali is going to reject login to users that aren't authorized any namespace.
````
```suggestion
Kiali is going to reject login to users that aren't authorized to see any namespace.
```

israel-hdez replied:
Fixed.

israel-hdez committed:
Add section for write privileges

Also:
* Fix review comments from Mazz
* Add OpenShift-related privileges.
  * In consequence, remove `kubectl` commands.

israel-hdez committed:
Change wording of pages that cross-link rbac.md

Also, name rbac.md as "Namespace access control" rather than "Namespace
autorizaton" or "RBAC".

israel-hdez commented:
This should be ready for review.

Preview link to the reworked page: https://deploy-preview-631--kiali.netlify.app/docs/configuration/rbac/

jmazzitelli commented:
Documenting the actual Role and all its permissions could be prone to "rot" (i.e. as we change the permissions in the code base in the future, this documentation would have to change, in fact depending on the version of Kiali you are using, these docs would be different). I guess we could ask the user to just look at the versioned kiali.io docs for the different versions of the Roles, but what if we did this instead? In the docs tell them to generate the ClusterRole from the server helm-chart - this generates a ClusterRole. If they want a per-namespace Role, they can edit the resulting yaml and change the Kind from ClusterRole to Role. For example, to generate a Kiali v1.57 Server ClusterRole with view_only permissions, you can do this:

```
helm template --version 1.57.0 --set deployment.view_only_mode=true --show-only templates/role-viewer.yaml kiali-server kiali/kiali-server
```

This results in:

```yaml
---
# Source: kiali-server/templates/role-viewer.yaml
apiVersion: rbac.authorization.k8s.io/v1
kind: ClusterRole
metadata:
  name: kiali-viewer
  labels:
    helm.sh/chart: kiali-server-1.57.0
    app: kiali
    app.kubernetes.io/name: kiali
    app.kubernetes.io/instance: kiali
    version: "v1.57.0"
    app.kubernetes.io/version: "v1.57.0"
    app.kubernetes.io/managed-by: Helm
    app.kubernetes.io/part-of: "kiali"
rules:
- apiGroups: [""]
  resources:
  - configmaps
  - endpoints
  - pods/log
  verbs:
  - get
  - list
  - watch
- apiGroups: [""]
  resources:
  - namespaces
  - pods
  - replicationcontrollers
  - services
  verbs:
  - get
  - list
  - watch
- apiGroups: [""]
  resources:
  - pods/portforward
  verbs:
  - create
  - post
- apiGroups: ["extensions", "apps"]
  resources:
  - daemonsets
  - deployments
  - replicasets
  - statefulsets
  verbs:
  - get
  - list
  - watch
- apiGroups: ["batch"]
  resources:
  - cronjobs
  - jobs
  verbs:
  - get
  - list
  - watch
- apiGroups:
  - networking.istio.io
  - security.istio.io
  - extensions.istio.io
  - telemetry.istio.io
  - gateway.networking.k8s.io
  resources: ["*"]
  verbs:
  - get
  - list
  - watch
- apiGroups: ["apps.openshift.io"]
  resources:
  - deploymentconfigs
  verbs:
  - get
  - list
  - watch
- apiGroups: ["project.openshift.io"]
  resources:
  - projects
  verbs:
  - get
- apiGroups: ["route.openshift.io"]
  resources:
  - routes
  verbs:
  - get
- apiGroups: ["authentication.k8s.io"]
  resources:
  - tokenreviews
  verbs:
  - create
...
```

If you want the read-write permissions, you can do:

```
helm template --version 1.57.0 --show-only templates/role.yaml kiali-server kiali/kiali-server
```

So we wouldn't have to change the docs as we change the Roles, we can leave just that helm template command and the user can generate the yaml that they want, for whatever version they want.

israel-hdez commented:
> If you want the read-write permissions, you can do:
> 
> ```
> helm template --version 1.57.0 --show-only templates/role.yaml kiali-server kiali/kiali-server
> ```


The problem with that, and the motivation for being explicit in the docs, is that you don't need all those privileges. As mentioned in the new writing, you just need this minimal privs:

```yaml
apiVersion: rbac.authorization.k8s.io/v1
kind: ClusterRole
metadata:
  name: kiali-namespace-authorization
rules:
- apiGroups: [""]
  resources:
  - namespaces
  - pods/log
  verbs:
  - get
- apiGroups: ["project.openshift.io"] # Only if you are using OpenShift
  resources:
  - projects
  verbs:
  - get
```

With this minimal role, Kiali will let you view everything else.

Looking forward, since the kube cache is now mandatory in Kiali, I think it is expected that this minimal role will remain stable, even if the KialiSA needs to get more privileges.

This is true for **read** privileges.

For **write** privileges, you _do_ need to give explicit privs, but the helm charts do not currently have a way to generate a Role or ClusterRole only for write privs.

israel-hdez commented:
What I want to communicate in the new docs, is that **you can't** fine-tune read privileges in Kiali. If you can do `kubectl get namespace foo`, Kiali will let you read everything in that namespace even if you don't have any other privilege. So, it doesn't make sense to take the KialiSA role as a template.

jmazzitelli commented:
@israel-hdez OK, I think I get it now. You are documenting the USER'S Role which doesn't need all the read-permissions the Kiali SA has (essentially, you are saying the USER will get the read-permissions that the Kiali SA has, regardless of what the USER'S read permissions are). Correct?

israel-hdez commented:
> (essentially, you are saying the USER will get the read-permissions that the Kiali SA has, regardless of what the USER'S read permissions are). Correct?

Yes! "So-so". Always limited to what the Kiali API lets you get -- e.g The Kiali SA may have privs to read some secrets, but Kiali does not have an API that would let you read them :-)

jmazzitelli commented on content/en/docs/Configuration/authentication/openid.md:
```diff
@@ -14,9 +14,8 @@ third-party system.
-[Kubernetes authorization mechanisms](https://kubernetes.io/docs/reference/access-authn-authz/rbac/). See the
-[RBAC documentation]({{< relref "../rbac" >}}) for more details.
+then Kiali's `openid` strategy can offer 
+[namespace access control]({{< relref "../rbac" >}}) for more details.
```
```suggestion
[namespace access control]({{< relref "../rbac" >}}).
```

jmazzitelli commented on content/en/docs/Configuration/rbac.md:
```diff
@@ -1,120 +1,214 @@
 
-Kiali supports role-based access control (RBAC) when using any authentication strategy other than `anonymous`.
+In authentication strategies other than `anonymous` Kiali supports limiting the
+namespaces that are accessible in a per-user basis. The `anonymous`
```
```suggestion
namespaces that are accessible on a per-user basis. The `anonymous`
```

jmazzitelli commented on content/en/docs/Configuration/rbac.md:
```diff
@@ -1,120 +1,214 @@
 
-## Minimum required privileges to login
+In general, Kiali will give _read_ access to namespaces where the logged in
+user is granted to _"GET"_ its definition -- i.e. the user is allowed to do a
```
```suggestion
user is allowed to _"GET"_ its definition -- i.e. the user is allowed to do a
```

jmazzitelli commented on content/en/docs/Configuration/rbac.md:
```diff
@@ -1,120 +1,214 @@
+Changing resources in the cluster can be a sensitive operation. Because of
+this, the logged in user will need to be given the needed privileges to perform
+any updates through Kiali. The following `ClusterRole` contains all write
+privileges that may used in Kiali:
```
```suggestion
privileges that may be used in Kiali:
```

jmazzitelli commented on content/en/docs/Configuration/rbac.md:
```diff
@@ -1,120 +1,214 @@
-`ClusterRoles` because they are bound to the Kiali ServiceAccount and editing
-them may lead to Kiali not working properly.
+{{% alert color="info" %}}
+If needed, you can reduce the set of write privileges to prevent users changing
```
```suggestion
If needed, you can reduce the set of write privileges to prevent users from changing
```

jmazzitelli reviewed (changes requested):
just some minor things

israel-hdez committed:
Fix Mazz feedback

Co-authored-by: John Mazzitelli <mazz@redhat.com>

israel-hdez commented:
@jmazzitelli I applied all your suggestions.
